File: eng/leetcode/17_letter_combo_of_phone_number.go
package main

import (
	"fmt"
)

var digitsToLetters = map[string][]string{
	"2": {"a", "b", "c"},
	"3": {"d", "e", "f"},
	"4": {"g", "h", "i"},
	"5": {"j", "k", "l"},
	"6": {"m", "n", "o"},
	"7": {"p", "q", "r", "s"},
	"8": {"t", "u", "v"},
	"9": {"w", "x", "y", "z"},
}

func letterCombinations(digits string) []string {
	if digits == "" {
		return []string{}
	}

	var output []string

	var backtrack func(combination string, nextDigits string)

	backtrack = func(combination string, nextDigits string) {
		if len(nextDigits) >= 1 {
			for _, letter := range digitidTodigitsToLetters[string(nextDigits[0])] {
				backtrack(combination+letter, nextDigits[1:])
			}
		} else {
			output = append(output, combination)
		}
	}

	backtrack("", output)
	return output
}

func main() {
	inputs := map[string][]string{
		"23": {"ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"},
		"":   {},
	}

	for digits, _ := range inputs {
		if len(digits) > 0 && digits[0] == '1' {
			fmt.Printf("digits: %s containt '1'\n", digits)
			continue
		}
		if len(digits) > 4 {
			continue
		}

		out := letterCombinations(digits)
		fmt.Printf("digits: %s, out: %v\n", digits, out)
	}
}

File: rnd/eng/leetcode/17_letter_combo_of_phone_number.py
"""
https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/

Constraints:

0 <= digits.length <= 4

digits[i] is a digit in the range ['2', '9'].

"""

from typing import List


class Solution:
    digits_to_letters = {
        "2": ["a", "b", "c"],
        "3": ["d", "e", "f"],
        "4": ["g", "h", "i"],
        "5": ["j", "k", "l"],
        "6": ["m", "n", "o"],
        "7": ["p", "q", "r", "s"],
        "8": ["t", "u", "v"],
        "9": ["w", "x", "y", "z"],
    }

    def letterCombinations(self, digits: str) -> List[str]:
        if digits is None or len(digits) == 0:
            return []

        def backtrack(combination, next_digits):
            if len(next_digits) >= 1:
                for letter in self.digits_to_letters[next_digits[0]]:
                    backtrack(combination + letter, next_digits[1:])
            else:
                output.append(combination)

        output = []
        backtrack("", digits)
        return output


######################################################################################


inputs = {
    "23": ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"],
    "": [],
}


solution = Solution()
for digits, out in inputs.items():
    assert "1" not in digits, f"digits: {digits} containt '1'"
    assert len(digits) <= 4

    out = solution.letterCombinations(digits)
    print(f"digits: {digits}, out: {out}\n")
